emmarin/cl/cl/src/crust/tx.rs

```diff
@@ -0,0 +1,334 @@
+use rand_core::{CryptoRngCore, RngCore};
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+
+use crate::{
+    crust::{
+        balance::{Balance, BalanceWitness},
+        iow::{BurnWitness, InputWitness, MintWitness, OutputWitness},
+        NoteCommitment, Nullifier,
+    },
+    ds::mmr::{MMRProof, MMR},
+    mantle::ZoneId,
+    merkle, Digest, Hash,
+};
+
+pub const MAX_INPUTS: usize = 8;
+pub const MAX_OUTPUTS: usize = 8;
+
+/// An identifier of a transaction
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
+pub struct TxRoot(pub [u8; 32]);
+
+/// An identifier of a bundle
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
+pub struct BundleRoot(pub [u8; 32]);
+
+impl From<[u8; 32]> for TxRoot {
+    fn from(bytes: [u8; 32]) -> Self {
+        Self(bytes)
+    }
+}
+
+impl From<[u8; 32]> for BundleRoot {
+    fn from(bytes: [u8; 32]) -> Self {
+        Self(bytes)
+    }
+}
+
+impl TxRoot {
+    pub fn random(mut rng: impl RngCore) -> Self {
+        let mut sk = [0u8; 32];
+        rng.fill_bytes(&mut sk);
+        Self(sk)
+    }
+
+    pub fn hex(&self) -> String {
+        hex::encode(self.0)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Tx {
+    pub root: TxRoot,
+    pub balance: Balance,
+    pub updates: Vec<LedgerUpdate>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct TxWitness {
+    pub inputs: Vec<InputWitness>,
+    pub outputs: Vec<(OutputWitness, Vec<u8>)>,
+    pub data: Vec<u8>,
+    pub balance_blinding: [u8; 16],
+    pub mint_burn_blinding: [u8; 16],
+    pub mints: Vec<MintWitness>,
+    pub burns: Vec<BurnWitness>,
+    pub frontier_paths: Vec<(MMR, MMRProof)>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct LedgerUpdate {
+    pub zone_id: ZoneId,
+    pub frontier_nodes: Vec<MMR>,
```

Comment:
This is actually a `Vec<Roots>`, the point was to de-dupe common roots across multiple MMR's.

Maybe we can simplify this down to one MMR per zone?

```
pub struct LedgerUpdate {
    pub zone_id: ZoneId,
    pub zone_mmr: MMR,
Comment
    pub inputs: Vec<Nullifier>,
    pub outputs: Vec<NoteCommitment>,
}
```

Reply:
I've put a comment about de-duping roots but I was lazy. I think it's reasonable to request this for a tx but less so for a bundle, so we might still need de-duping there

Reply:
yes, deduping in the bundle is already implemented IIRC

Reply:
I don't think we do actually, but I'll add it

Reply:
ah ok, I believe we used to

Reply:
readded it